mitodl/ocw-hugo-themes
# feat: add off-site navigation confirmation dialog

# What are the relevant tickets?
Closes https://github.com/mitodl/hq/issues/1812 

# Description (What does it do?)

This PR:

- Adds an icon next to external sidebar links to distinguish them.
- Adds a confirmation dialog that opens before the page navigates to the external link.

## Choice of approach

A single modal's markdown is placed inside the `body` tag. The confirm action link is dynamically updated when a user clicks an external link.

Bootstrap's modals are sensitive to positioned parents. This is the reason why the modal needs to be placed inside the `body` tag.

## Design: The off-bits

You might observe that the output looks a bit different from the [designs](https://www.figma.com/proto/7TIcfINO8O8SPVyALe632H/File-Size?page-id=0%3A1&type=design&node-id=165-2&viewport=122%2C-1406%2C0.62&t=Z28hyDE4MOM5HpBE-1&scaling=min-zoom&starting-point-node-id=165%3A2&mode=design) (which might not be obvious). The reasons for this are:

* The design uses `Arial` while we use `Helvetica` font.
* The side nav link icons are aligned to the baseline of the text, which is in accordance with the design. But the design has no [descenders](https://mockitt.wondershare.com/dictionary/what-is-baseline.html).
* I've reused our existing font measurements, instead of creating new ones, to avoid tech debt later on. There is a slight, but almost negligible, difference in those font sizes.

The following image shows a stress test of the icon placement design.
<img width="238" alt="Screenshot 2023-10-23 at 1 28 08 PM" src="https://github.com/mitodl/ocw-hugo-themes/assets/71316217/d954791e-42e7-4316-96bb-5d79b9b4a541">


# Screenshots (if appropriate):
<img width="1680" alt="Screenshot 2023-10-23 at 2 53 02 PM" src="https://github.com/mitodl/ocw-hugo-themes/assets/71316217/0c9d4847-225b-4b51-99d9-a52ad3ac1949">

<img width="500" alt="Screenshot 2023-10-23 at 2 53 14 PM" src="https://github.com/mitodl/ocw-hugo-themes/assets/71316217/69b79884-0d60-483e-88f0-4ef683a8248c">


https://github.com/mitodl/ocw-hugo-themes/assets/71316217/bfc87c2c-2ea7-4ab6-ab0f-9141c11be6a2

https://github.com/mitodl/ocw-hugo-themes/assets/71316217/82ec1ab8-dd69-4a01-9aa4-c676152f5535

# How can this be tested?

1. Grab content for a course from [this list](https://github.com/mitodl/hq/issues/1812#issuecomment-1745509175).
2. Place the content in your local content directory.
3. Navigate to your local setup of ocw-hugo-themes.
4. Checkout branch `hussaintaj/1812-external-link-modal`.
5. Run
    ```
    yarn start course <course-id>
    ```
6. Observe the external links in the sidebar.
    * Expect external links to be marked with an icon.
    * Expect a dialog to pop-up when you click on an external link.
7. Observe the confirmation dialog.
    * Expect that clicking "Stay Here", the close button, or outside the modal closes the dialog.
    * Expect that clicking on "Continue" takes you to the external link.
